File: src/services/token-manager.ts
export class TokenManager {
  private totalTokens = 0;
  private lastShownThreshold = 0;

  updateUsage(
    input: number,
    output: number,
    cacheCreation = 0,
    cacheRead = 0,
  ): void {
    // input_tokens alone does NOT include cache tokens or output tokens.
    // The actual context window usage is:
    // input_tokens + output_tokens + cache_creation_input_tokens + cache_read_input_tokens
    this.totalTokens = input + output + cacheCreation + cacheRead;
  }

  getTotal(): number {
    return this.totalTokens;
  }

  getRatio(contextLength: number): number {
    return this.totalTokens / contextLength;
  }

  shouldCompress(contextLength: number, thresholdRatio: number): boolean {
    const threshold = Math.floor(contextLength * thresholdRatio);
    return this.totalTokens > threshold;
  }

  getLastShownThreshold(): number {
    return this.lastShownThreshold;
  }

  updateThreshold(value: number): void {
    this.lastShownThreshold = value;
  }

  reset(): void {
    this.totalTokens = 0;
    this.lastShownThreshold = 0;
  }
}

File: internal/token_manager.go
package internal

// TokenManager tracks token usage and compression thresholds.
type TokenManager struct {
	totalTokens        int
	lastShownThreshold int
}

// NewTokenManager creates a new token manager.
func NewTokenManager() *TokenManager { return &TokenManager{} }

// Add updates the total token count. Input includes cache tokens.
func (tm *TokenManager) Add(input, output, cacheCreation, cacheRead int) {
	tm.totalTokens = input + output + cacheCreation + cacheRead
}

// Total returns the total token count.
func (tm *TokenManager) Total() int { return tm.totalTokens }

// Ratio returns the fraction of context used.
func (tm *TokenManager) Ratio(contextLength int) float64 {
	if contextLength == 0 {
		return 0
	}
	return float64(tm.totalTokens) / float64(contextLength)
}

// ShouldCompress returns true when tokens exceed threshold ratio.
func (tm *TokenManager) ShouldCompress(contextLength int, thresholdRatio float64) bool {
	threshold := int(float64(contextLength) * thresholdRatio)
	return tm.totalTokens > threshold
}

// LastShownThreshold returns the last recorded percentage threshold.
func (tm *TokenManager) LastShownThreshold() int { return tm.lastShownThreshold }

// SetLastShownThreshold updates the last shown threshold.
func (tm *TokenManager) SetLastShownThreshold(v int) { tm.lastShownThreshold = v }

// Reset clears all counters.
func (tm *TokenManager) Reset() {
	tm.totalTokens = 0
	tm.lastShownThreshold = 0
}
